rat/ops.py

```diff
@@ -108,15 +250,111 @@ def scalar(self):
     def get_key(self):
         return self.name
 
-    def code(self):
+    def fold(self):
+        if self.lower is not None:
+            self.lower = self.lower.fold()
+            if not isinstance(self.lower, (RealConstant, IntegerConstant)):
+                raise ConstantFoldError(
+                    f"Lower bound value must fold into a Numeric constant at compile time, but folded expression {self.lower} is not a constant!",
+                    self.line_index,
+                    self.column_index,
+                )
+
+        if self.upper is not None:
+            self.upper = self.upper.fold()
+            if not isinstance(self.upper, (RealConstant, IntegerConstant)):
+                raise ConstantFoldError(
+                    f"Upper bound value must fold into a Numeric constant at compile time, but folded expression {self.upper} is not a constant!",
+                    self.line_index,
+                    self.column_index,
+                )
+
+        if self.subscript:
+            self.subscript = self.subscript.fold()
+
+        return Param(
+            name=self.name,
+            prime=self.prime,
+            subscript=self.subscript,
+            lower=self.lower,
+            upper=self.upper,
+            assigned_by_scan=self.assigned_by_scan,
+            line_index=self.line_index,
+            column_index=self.column_index,
+        )
+
+    def __post_init__(self):
+        self.out_type = types.NumericType
+
+    def code(self, scalar=False):
         variable_code = self.variable.code()
         if self.subscript:
-            return f"parameters['{variable_code}'][{self.subscript.code()}]"
+            if self.assigned_by_scan:
+                if scalar == False:
+                    raise Exception("Internal Error: Cannot do a vector evaluation of an assigned variable")
+
+                return f"carry[index - {self.subscript.code(scalar)} - 1]"
+            else:
+                return f"parameters['{variable_code}'][{self.subscript.code(scalar)}]"
         else:
             return f"parameters['{variable_code}']"
 
     def __str__(self):
-        return f"Param({self.name}, {self.subscript.__str__()}, lower={self.lower}, upper={self.upper})"
+        return f"Param({self.name}, subscript={self.subscript.__str__()}, lower={self.lower}, upper={self.upper}, assigned={self.assigned_by_scan})"
+
+
+@dataclass
+class Match(Expr):
+    first: Expr
+    otherwise: Expr
```

Comment:
When I moved match to be an expression I changed it to be more like an if-else just to make things simple on myself. If we pull out match, I'll just move all the stuff to a different PR so we have it in the future.